lightwood/encoders/text/pretrained.py

```diff
@@ -0,0 +1,323 @@
+            )
+
+        else:
+            print("Embeddings Generator only", flush=True)
+
+            self.model_type = "embeddings_generator"
+            self._model = self._embeddings_model_class.from_pretrained(
+                self._pretrained_model_name
+            ).to(self.device)
+
+        self._prepared = True
+
+    def _train_model(self, dataset, optim=None, scheduler=None, n_epochs=1):
+        """
+        Given a model, train for n_epochs.
+
+        model - torch.nn model;
+        dataset - torch.DataLoader; dataset to train
+        device - torch.device; cuda/cpu
+        log - lightwood.logger.log; print output
+        optim - transformers.optimization.AdamW; optimizer
+        n_epochs - number of epochs to train
+
+        """
+        self._model.train()
+
+        if optim is None:
+            print("No opt. provided, setting all params with AdamW.", flush=True)
+            optim = AdamW(self._model.parameters(), lr=5e-5)
+        else:
+            print("Optimizer provided", flush=True)
+
+        if scheduler is None:
+            print("No scheduler provided.", flush=True)
+        else:
+            print("Scheduler provided.", flush=True)
+
+        print("Beginning training.")
+        for epoch in range(n_epochs):
+            total_loss = 0
+
+            for batch in dataset:
+                optim.zero_grad()
+
+                with LightwoodAutocast():
+                    inpids = batch["input_ids"].to(self.device)
+                    attn = batch["attention_mask"].to(self.device)
+                    labels = batch["labels"].to(self.device)
+                    outputs = self._model(inpids, attention_mask=attn, labels=labels)
+                    loss = outputs[0]
+
+                total_loss += loss.item()
+
+                loss.backward()
+                optim.step()
+                if scheduler is not None:
+                    scheduler.step()
+
+            # self._train_callback(epoch, loss.item())
+            print("Epoch=", epoch + 1, "Loss=", total_loss / len(dataset), flush=True)
+
+    def _train_callback(self, epoch, loss):
+        log.info(f"{self.name} at epoch {epoch+1} and loss {loss}!")
+
+    def encode(self, column_data):
+        """
+        TODO: Maybe batch the text up; may take too long
+        Given column data, encode the dataset.
+
+        Currently, returns the embedding of the pre-classifier layer.
+
+        Args:
+        column_data:: [list[str]] list of text data in str form
+
+        Returns:
+        encoded_representation:: [torch.Tensor] N_sentences x Nembed_dim
+        """
+        if self._prepared is False:
+            raise Exception("You need to first prepare the encoder.")
+
+        # Freeze training mode while encoding
+        self._model.eval()
+
+        encoded_representation = []
+
+        with torch.no_grad():
+            # Set the weights; this is GPT-2
+            for text in column_data:
```

Comment:
Batching here, even if in pairs of 5 or 10 due to size, might actually be *very* useful, especially since with text datasets we might not afford to store the embeddings in full (due to size, i.e. we'll have to disable the cache or it will get auto-disabled) leading to this function being called very often.

But I'm like 40% sure this would be useful, if it's too annoying don't do it.

Reply:
Can do- tokenizing for sure can be done in batching.